### PR TITLE
node: handle failed quorums to produce attested block

### LINKED ISSUE #1268: consensus: Implement new Validation and Ratification logic
#### Summary

The current implementation of the Validation and Ratification steps is not aligned with the intended design of #1177.
Additionally, the original design does not align with slashing, which has been later introduced with #1257. 

This Issue describes in detail how the Validation and Ratification steps should work in the new protocol.

#### Detailed Description

Nomenclature Update:
- We will use _quorum_ to refer to two quantities:
  - The _supermajority_ 2/3 
  - The simple _majority_ 1/2+1
- _Target  Quorum_ will refer to the quorum threshold depending on the type of vote 
  
- NoQuorum: represent the vote casted if the amount of votes is insufficient to reach a conclusion on the step; it is only casted in the Ratification step

**VALIDATION**
- Votes:
  - _Valid_: a candidate block has been received and it's valid
  - _Invalid_: a candidate block has been received but it's invalid
  - _NoCandidate:_ no candidate block has been received

- Validation message:
  - Header:
     - Prev_Hash
     - Round
     - Iteration
     - StepNum (= "Validation")
  -  Payload:
     - _Vote_: Valid, Invalid, NoCandidate
     - _Candidate_Hash_: empty if Vote=NoCandidate
     - _Signature_: Sign(Hash(Header || Vote || Candidate_Hash))

- ValidationResults:
 - _Valid_: a Supermajority of Valid votes has been received
 - _Invalid_: a Majority of Invalid votes has been received
 - _NoCandidate_: a Majority of NoCandidate votes has been received
 - _NoQuorum_: the timeout expired 

**RATIFICATION**
The Ratification vote corresponds to the local Validation Result and is accompanied by the SV of the Validation votes that produced the result.

- Votes
  - _Valid_
  - _Invalid_
  - _NoCandidate_
  - _NoQuorum_

- Ratification message:
  - Header:
     - Prev_Hash
     - Round
     - Iteration
     - StepNum (= "Ratification")
  -  Payload:
     - Vote: Valid, Invalid, NoCandidate, NoQuorum
     - Candidate_Hash: empty if Vote=NoCandidate or Vote=NoQuorum
     - Signature: Sign(Hash(Header || Vote || Candidate_Hash))
     - Validation_Votes: StepVotes of the Validation votes that determined the Ratification Vote; empty if Vote=NoQuorum

Each Ratification vote is verified against Validation_Votes. Therefore, a NoCandidate vote is only valid with a SV containing a MajorityQuorum of NoCandidate votes for Validation.

- Ratification Results:
  - Success (Valid): a Supermajority Quorum of Valid votes has been received -> broadcast Quorum message
  - Fail (Invalid): a Majority Quorum of Invalid votes has been received -> broadcast Quorum message
  - Fail (NoCandidate): a Majority Quorum of NoCandidate votes has been received -> broadcast Quorum message
  - Fail(NoQuorum): a Majority Quorum of NoQuorum votes has been received -> broadcast Quorum message
  - Timeout:  -> no output

Quorum messages with Fail results are used to attest failed iterations. Invalid and NoCandidate fails are additionally used for slashing.
  
- Quorum message:
  - Header:
     - Prev_Hash
     - Round
     - Iteration
     - StepNum: not necessary
  -  Payload:
     - Result: Success, Fail (Invalid), Fail (NoCandidate), Fail(NoQuorum)
     - Candidate_Hash: empty if Vote=NoCandidate or Vote=NoQuorum
     - Validation_Votes
     - Ratification_Votes

Votes are checked against the Result's vote (Valid,Invalid,NoCandidate,NoQuorum), which must be the same vote for both Validation and Ratification. An exception to this is NoQuorum, which can differ from the Validation result.

**ProcessQuorum**
- Success: accept candidate
- Fail (Invalid): add to FailedIterations; slash for invalid block
- Fail (NoCandidate): add to FailedIterations; slash for missed block
- Fail (NoQuorum): add to FailedIterations
  
#### Relevant Context
 - Waiting for a majority of Invalid/NoCandidate votes in Validation allows to mutually exclude these votes in the Ratification step (since there can't be a Ratification vote NoCandidate if there was a majority of Invalid votes in Validation)
 -  In Ratification:
   - we wait for a Supermajority of Valid in order to be sure most provisioners are accepting the new block (i.e., they are aligned with the other provisioners)
   - we wait for a Majority of Invalid/NoCandidate for the same reason as for Valid: we want a large portion of provisioners to be aligned on the iteration as failed.
 - While a single NoQuorum result on Validation is not a definitive result (there might have been a Quorum seen by other provisioners), a majority of NoQuorum in the Ratification step ensure no other result is possible, hence we produce the Quorum message that can be used to attest the iteration as failed.

 #### Implementation
  - Consensus messages are being updated by #1264 

#### Notes
  - $NoCandidate$ and $NoQuorum$ votes should be disabled in Emergency Mode, since step timeouts do not interrupt the step itself.

### PR DESCRIPTION
**Pull request by herr-seppia**
Resolves #1268

### CODE REVIEW

File: node/src/chain/header_validation.rs [hunk 1]
```diff
@@ -147,24 +145,21 @@ impl<'a, DB: database::DB> Validator<'a, DB> {
...
                     candidate_block.height,
                     cert,
                     iter as u8,
                 )
                 .await?;
 
-                attested = attested
-                    && quorums.0.quorum_reached()
-                    && quorums.1.quorum_reached();
+                // Ratification quorum is enough to
```
Comment (fed-franz) [inline node/src/chain/header_validation.rs:161]: is enough to... ?

File: node/src/chain/header_validation.rs [hunk 2]
```diff
@@ -147,24 +145,21 @@ impl<'a, DB: database::DB> Validator<'a, DB> {
...
-                    anyhow::bail!("Invalid generator. Expected {expected_pk:?}, actual {pk:?}");
-                }
+
+                anyhow::ensure!(pk == &expected_pk, "Invalid generator. Expected {expected_pk:?}, actual {pk:?}");
 
                 let quorums = verify_block_cert(
                     self.prev_header.hash,
                     self.prev_header.seed,
                     self.provisioners.current(),
-                    Vote::NoCandidate,
```
Comment (fed-franz) [inline node/src/chain/header_validation.rs:158]: Not sure if this is done somewhere else, but we should verify the result is Fail

File: consensus/src/step_votes_reg.rs [hunk 3]
```diff
@@ -97,14 +97,24 @@ impl CertificateInfo {
     /// Returns `true` if all fields are non-empty and quorum is reached for
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:97]: This should be updated

File: consensus/src/step_votes_reg.rs [hunk 4]
```diff
@@ -199,9 +215,7 @@ impl CertInfoRegistry {
                 self.cert_list
                     .get(&iteration)
                     .and_then(|iter| {
-                        iter.votes
-                            .get(&Vote::NoCandidate)
-                            .map(|ci| (ci, iter.generator))
+                        iter.failed().map(|ci| (ci, iter.generator))
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:218]: `get_nil_certificates` should be renamed to `get_failed_certificates`

**Review by fed-franz — CHANGES_REQUESTED** [review]
A couple of nits, and I think we should check failed certificates are indeed for a Fail result.
The rest LGTM

On a side note, not sure what's the standard procedure, but this PR only partially resolves the issue. 
Maybe the description should state that.

File: node/src/chain/header_validation.rs (hunk 2 above)
Comment (herr-seppia) [reply node/src/chain/header_validation.rs:158]: you're right, added

**Review by fed-franz — APPROVED** [review]

**Pull request merged**